### ISSUE TITLE
a bug in relational storage reminders implementation

### ISSUE THREAD

**Issue by shayhatsor** [dotnet/orleans#1365]
we might have a problem with the current implementation of reminders in relational storage
the code casts the start and end grain hash from uint to int before querying...
i discovered it because i wrote a new test:
```csharp
//after adding some reminders
var rows = await reminderTable.ReadRows(0, uint.MaxValue);
```
you'd expected to get all the reminders, right ?
well... it depends if you're lucky...
what happens is that if the hash is above 2^31-1 the value that's saved in the db is negative...
now this query:
```sql
'ReadRangeRows1Key','
    SELECT
        GrainId,
        ReminderName,
        StartTime,
        Period,
        ETag
    FROM OrleansRemindersTable
    WHERE
        ServiceId = @serviceId AND @serviceId IS NOT NULL
        AND GrainIdConsistentHash > @beginHash AND @beginHash IS NOT NULL
        AND GrainIdConsistentHash <= @endHash AND @endHash IS NOT NULL;
```
will never return it...

```sql
'ReadRangeRows2Key','
    SELECT
        GrainId,
        ReminderName,
        StartTime,
        Period,
        ETag
    FROM OrleansRemindersTable
    WHERE
        ServiceId = @serviceId AND @serviceId IS NOT NULL
        AND ((GrainIdConsistentHash > @beginHash AND @beginHash IS NOT NULL)
        OR (GrainIdConsistentHash <= @endHash AND @endHash IS NOT NULL));
```
will return it, because it returns all negative hashes in our case.
so best case we got some really bad load balancing and one silo is handling half of the reminders,
and in the worst case we might never fire **half** of the reminders.
it seems that the ring's "last" silo's range can have `beginhash`>`endhash`. this silo will get all the negatives.
it's an easy fix, but any system that's currently in production using `MSSQL` or `MySQL` will need a migration script. and we'd probably need to use `BIGINT` in `MSSQL` because it doesn't have unsigned numbers :astonished:

**Comment by sergeybykov** [comment]
>Engineer walks into a bar. Orders a beer. Orders 0 beers. Orders MAXINT of beers. Orders a lizard. Orders -1 beers. Orders a sfdeljknesv.

Good catch. :-)

**Comment by gabikliot** [comment]
We had similar issues with Azure based reminders in the past, need to be very careful about those things. We are now very explicit about uint: https://github.com/dotnet/orleans/blob/master/src/OrleansAzureUtils/Storage/RemindersTableManager.cs#L117

It complies, runs fine, and passes all tests. The "failure" is silent and only at scale. Very hard to find. The only way I see those things can be prevented is via directed white box (unit) testing, where you assert the behavior directly.

**Comment by shayhatsor** [comment]
I'm pretty sure the bug existed since day one, since `MSSQL` doesn't support unsigned int. The question is what should we do now ? obviously fix it. but what about people that have DBs full of reminders ? the thing is, the data is there; it's just not in the *right* format. we can prepare a migration script, but then it might create a huge burst of reminders and crash clusters.

**Comment by veikkoeeva** [comment]
@shayhatsor Well spotted. Looking [at the history](https://github.com/dotnet/orleans/commits/17e52c9f08ab30b0cc8c204e2ba7cb08822753f8/src/OrleansProviders/SQLServer/CreateTables.sql), this has actually been as you describe as a fix, but got changed around Orleans became public. Very unfortunate no one catched it before, well spotted. It makes me wonder if all the customers that run Orleans have updated themselves and should they be accounted for in the migration script?

How to do versioning: Should we include something like `CreateOrleansTables_fromversion_to_version_vendor.sql`, where vendor is optional if it cannot be made in a cross-platform way?

I had the idea to use [OrleansDatabaseInfo](https://github.com/dotnet/orleans/blob/master/src/OrleansSQLUtils/CreateOrleansTables_SqlServer.sql) to include the version number of the schema and *maybe* in addition to versioning individual tables. This way one could check the version number of the schema and use that to run `ALTER` statements. It should be easier than checking arbitrary conditions. I believe having a version table or something is a common approach (I'm not that a seasoned DB developer/DBA, so take with a pinch of salt). That could be even printed upon start-up to make troubleshooting easier (on the forums). What do you think?

A slight digression, I note that many vendors include versioning functionality, but I believe even then they aren't always used and they might become a hassle to do a cross-platform way. For SQL Server, if one has wrapped the script into a [dacpack deployment](https://msdn.microsoft.com/en-us/library/ee210546.aspx), it'd look like [this](http://stackoverflow.com/questions/3331782/how-do-i-find-the-current-version-of-a-sql-server-data-tier-application) and I believe it can handle this conversion (semi-)automatically.

**Comment by veikkoeeva** [comment]
Moreover, looking at history, has storage been an issue before? Maybe with millions and millions of reminders? If so, @shayhatsor did optimize the MySQL script on this regard and the optimization hasn't been incorporated to SQL Server.

**Comment by shayhatsor** [comment]
@veikkoeeva, the problem here isn't about the quantity of reminders, it's about the distribution of grain hashes, which AFAIK is distributed evenly across the unsigned int. so you can have the bug with even one grain that just happened to be hashed to something higher than 2^31 - 1. which has a 50% chance of happening, so it's basically a coin toss :smile: 
in the current `MySQL` script there's the same problem, i did my best to *not* optimize anything there; in order to keep it as close to the `MSSQL` one as possible. i mentioned that `MSSQL` doesn't support unsigned numbers because i found it hard to believe someone changed the field from bigint to int, but someone actually did ! [here](https://github.com/dotnet/orleans/commit/a5dde6b6e193aea4ce5e299e3bdb69d3fd4783fa)

**Comment by shayhatsor** [comment]
@sergeybykov, we're in the spirit of quotes, so this comes to mind:
Premature optimization is the root of all evil -- `Donald Knuth`

**Comment by veikkoeeva** [comment]
@shayhatsor OK, I remember it wrong that you did something smart that reduced the storate required by reminders. This is certainly true for membership version data, but one hardly runs millions of silos! Otherwise I fully agree on the actual problem.

Fix is the issue now, as you pointed. At least it requires in the release notes:

* a description of this issue at hand
* the options one has
* the assumed effects of suggested fixes once applied.

I wonder if we have more choices here than to provide a simple upgrade script? It could convert some of the reminders in batches in order to prevent a reminder storm. Then again, it should be expected if there are a lot of reminders (albeit not encountered before applying this). One option is also to just fix the deployment scripts and point out that user interaction is required and let those who have deployed in production to do what best fits them (maybe using tooling they have, Dacpack, Red Gate, whatever).

**Comment by sergeybykov** [comment]
Not being a SQL expert, I presume we cannot easily change the column type transparently on the fly from within the `SqlReminderTable` code. If that's the case, then release notes about the fix seems like the only option we realistically have.

**Comment by veikkoeeva** [comment]
@sergeybykov  The column type can be changed even when running the application. It's `ALTER TABLE OrleansRemindersTable ALTER COLUMN GrainIdConsistentHash BIGINT NOT NULL;` for SQL Server. Just to give context, one can program relational engines like little operating systems. The language just looks more as LINQ and on should inherently care more about data and concurrency. :)

The other issue is that there should be a change in the code too, to put back [this](https://github.com/dotnet/orleans/commit/a5dde6b6e193aea4ce5e299e3bdb69d3fd4783fa) and, actually, in the code [here](private static IDbDataParameter CreateBeginHashParameter(IDbCommand command, uint beginHash)) these two methods should cast to `long` in order them to be mapped to [BIGINT](https://github.com/dotnet/orleans/blob/a64fa8953e764ac391ebd549d395ac66374b545f/src/OrleansSQLUtils/Storage/DbExtensions.cs#L26) (in MSDN [the mapping table](https://msdn.microsoft.com/en-us/library/cc716729.aspx)). (*)

In my unclearly phrased writing previously I meant that when the column type is changed and the queries all the sudden could find a lot of new reminders, which were previously skipped, could they consume enough resources in silos to bring them down? This shouldn't be a problem, I think. Also, one could have deployed the tables to a schema other than the default `dbo` (even likely) and maybe the best we can do is to tell in release notes to run the `ALTER TABLE` after the new binaries have been deployed.

(*) There's a discussion point that ADO.NET would convert parameters by itself, but in Orleans there's an explicit conversion table and it converts, say `uint`, so that SQL Server ADO.NET library balks an exception out of it. Should it be like so or should it silently convert to something else is the point, but not up for discussion here.

**Comment by shayhatsor** [comment]
@veikkoeeva, i'm working on a solution that will work with current db schema and data.
I've notices several things :
1) `GrainIdConsistentHash`'s data is there. converting from `uint` to `int` doesn't change the bits value.
2) the only problem is with the range queries.
3) when we query the table, the `GrainIdConsistentHash` parameter needs to be of the expected type (`uint` or `int`)

The solution:
* only once, get the `GrainIdConsistentHash`'s underlying .net type. i'm running a simple query to return the metadata using ADO.NET.
* when we need to run a reminders range query:
 * if the `GrainIdConsistentHash`'s underlying .net type is `int`, map `beginhash` and `endhash` to `int` values. This is done by converting to `long` and subtracting (`int.MaxValue`+1). then convert back to `int` and the ranges are now right.
*  if the `GrainIdConsistentHash`'s underlying .net type is `uint`, just pass it as is.
* when we upsert a reminder. send the `GrainIdConsistentHash` as is, but be sure to create the parameter with the right type (`uint` or `int`). here there's no mapping.

this change will allow an underlying db to use a `singed int` or `unsigned int` column types and Orleans will work right, transparently. so no need to change the scripts or current tables !

It works, but i'm still testing it.

**Comment by veikkoeeva** [comment]
@shayhatsor No rush as far as I'm concerned. Do sleep even as I'm in constant state of deprivation and can only see your idea does work and is a good one. :)

I can babble (though)... If we take it the hash fits into 32 bits, maybe it can be stored as signed and then converted using `Convert.ToUInt32(theHashFromTheDb)` and `Convert.ToInt32(theHashFromSilo)`. I suppose that only checks the sign bit, or "reinterprets it" (which, if this were C or C++, would be a place for dragons to roam due to things like a byte might not be eight bits, endianess and whatnot :)). If taking those conversions as the baseline, the issue looks like being if there needs to be done something for the conditional clauses?

**Comment by shayhatsor** [comment]
I think we don't need to do anything about the conditional clauses, because the range mapping covers all 2^32 possible values.

**Comment by gabikliot** [comment]
Fixed via #1374.
